File: Apps4Bro.Windows/AdManager.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.IO;
using System.Text;
using Apps4Bro.Networks;

namespace Apps4Bro
{
    public class AdManager
    {
        private readonly static string App4BroTag = "app4bro";

        private readonly Dictionary<string, AdNetworkHandler> m_adNetworks;
        private readonly Dictionary<string, string> m_keys;

        private readonly ReportManager m_reportManager;

        private AdWrapper[] m_adWrappers;
        private AdContextDelegate m_context;
        private int m_currentWrapper;
        private bool m_adShown;
        private bool m_adLoaded;
        private string m_appId;
        private object m_data;

        public bool IsAdShown
        {
            get { return m_adShown; }
        }

        public bool IsAdLoaded
        {
            get { return m_adLoaded; }
        }

        public bool IsInited
        {
            get { return m_adNetworks != null && m_adWrappers != null; }
        }
        public string AppId
        {
            get { return m_appId; }
        }

        public AdContextDelegate Context
        {
            get { return m_context; }
        }

        public ReportManager ReportManager
        {
            get { return m_reportManager; }
        }

        public AdManager(params string[] keys)
        {
            m_keys = new Dictionary<string, string>();

            for (int i = 0; i < keys.Length; i += 2)
                m_keys.Add(keys[i].ToLower(), keys[i + 1]);

            if (!m_keys.TryGetValue(App4BroTag, out m_appId))
                throw new ApplicationException("No App4Bro key specified for AdManager!");

            m_reportManager = new ReportManager(m_appId);

#if VERBOSE_LOG
            m_reportManager.ReportEvent("AD_START", Apps4BroSDK.Version.ToString());
#endif
            m_adNetworks = new Dictionary<string, AdNetworkHandler>();

            // Registering built-in SDKs
#if __IOS__

            // Interstitials
#if USE_ADMOB
            RegisterAdNetwork(new AdMobNetwork(this));
#endif
#if USE_ADCASH
            RegisterAdNetwork(new AdCashNetwork(this));
#endif
#if USE_IAD
            RegisterAdNetwork(new iAdNetwork(this));
#endif

            if (UIKit.UIDevice.CurrentDevice.CheckSystemVersion(7, 0))
            {
#if USE_APPODEAL
                RegisterAdNetwork(new AppodealNetwork(this));
#endif
#if USE_INNERACTIVE
                RegisterAdNetwork(new InneractiveNetwork(this));
#endif
            }

            // Banners
#if USE_ADMOB
            RegisterAdNetwork(new AdMobBannerNetwork(this));
#endif
#if USE_IAD
            RegisterAdNetwork(new iAdBannerNetwork(this));
#endif


            if (UIKit.UIDevice.CurrentDevice.CheckSystemVersion(7, 0))
            {
#if USE_INNERACTIVE
                RegisterAdNetwork(new InneractiveBanner(this));
#endif
#if USE_ADTOAPP
                RegisterAdNetwork(new AdToAppNetwork(this));
#endif
            }

#elif __ANDROID__
            RegisterAdNetwork(new AdMobNetwork(this));
#if USE_ADCASH
            RegisterAdNetwork(new AdCashNetwork(this));
#endif
#endif
            RegisterAdNetwork(new DummyNetwork(this));
            RegisterAdNetwork(new HouseNetwork(this));
        }

        public void RegisterAdNetwork(AdNetworkHandler handler)
        {
            m_adNetworks[handler.Network.ToLower()] = handler;
        }

        private string FormatRequest()
        {
            return string.Format(Apps4BroSDK.AdManagerUrl,
                Apps4BroSDK.AdvertisingId,
                m_appId
            );
        }

        public void Init(AdContextDelegate context, object data)
        {
            m_data = data;

            if (IsInited)
            {
                LoadAd();
                return;
            }

            m_context = context;
            WebRequest request = WebRequest.Create(FormatRequest());
            request.Headers[HttpRequestHeader.UserAgent] = "WinHTTP";
            request.ContentType = "application/text";
            request.Method = "GET";
#if !NETFX_CORE
            request.Timeout = 5000;
#endif

            request.BeginGetResponse(InitAsync, request);
        }

        private void InitAsync(IAsyncResult asynchronousResult)
        {
            try
            {
                string data = null;

                WebRequest request = (WebRequest)asynchronousResult.AsyncState;
                HttpWebResponse response = request.EndGetResponse(asynchronousResult) as HttpWebResponse;
                if (response == null || response.StatusCode != HttpStatusCode.OK)
                {
                    Debug.WriteLine("Error fetching data. Server returned status code: {0}", response == null ? HttpStatusCode.Unused : response.StatusCode);
                }
                else
                {
                    using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                        data = reader.ReadToEnd();
                }

                if (!string.IsNullOrWhiteSpace(data))
                {
                    ParseAdData(data, true);
                }
                else
                    Debug.WriteLine("Got empty response");

            }
            catch (Exception ex)
            {
                Debug.WriteLine("Network error: " + ex);
            }

            if (m_adWrappers == null) // try cached data
            {
#if __IOS__

                string data = Foundation.NSUserDefaults.StandardUserDefaults.StringForKey(App4BroTag + "_" + m_appId);
                if (data != null)
                {
                    Debug.WriteLine("Registering cached ads");
                    ParseAdData(data, false);
                }

#endif
            }
           

            if (m_adWrappers == null)
            {
                Debug.WriteLine("Registering fallback ads");

                StringBuilder builder = new StringBuilder();

                int count = 0;
                foreach (KeyValuePair<string, string> pair in m_keys)
                {
                    count++;
                    builder.AppendFormat("{0}:{0}def{1}|{2}|", pair.Key, count, pair.Value);
                }

                ParseAdData(builder.ToString(), false);

                if (m_adWrappers == null)
                {
                    Debug.WriteLine("No ad networks registered!");
                    return;
                }
            }

            m_currentWrapper = -1;

            m_context.OnInited(this);
        }

        private void ParseAdData(string data, bool save)
        {
            try
            {
                string[] adData = data.Split(new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries);

                if (adData.Length > 1)
                {
                    List<AdWrapper> wrappers = new List<AdWrapper>(adData.Length / 2);

                    for (int i = 0; i < adData.Length; i += 2)
                    {
                        string network = adData[i].ToLower();
                        string networkId;

                        if (network.Contains(":"))
                        {
                            string[] nsplit = network.Split(':');
                            network = nsplit[0];
                            networkId = nsplit[1];
                        }
                        else
                            networkId = network + " " + (wrappers.Count + 1);

                        AdNetworkHandler networkHandler = null;

                        if (m_adNetworks.TryGetValue(network, out networkHandler))
                        {
                            wrappers.Add(new AdWrapper(networkHandler, adData[i + 1], networkId));
                            Debug.WriteLine("Registered ad network " + network + "[" + adData[i + 1] + "]");
                        }
                        else
                            Debug.WriteLine("Failed to register ad network " + network);
                    }

                    if (wrappers.Count == 0)
                    {
                        Debug.WriteLine("No ad networks registered!");
                    }
                    else
                    {
                        m_adWrappers = wrappers.ToArray();
                        Debug.WriteLine("Registered " + m_adWrappers.Length + " ad networks");
                    }

                    if (save)
                    {
#if __IOS__
                        Foundation.NSUserDefaults.StandardUserDefaults.SetString(data, App4BroTag + "_" + m_appId);
#endif
                    }


                    m_reportManager.ReportEvent("AD_INIT", m_adWrappers.Length.ToString());
                }
            }
            catch (Exception)
            {
                Debug.WriteLine("Error parsing data string " + data);
            }
        }

        public void HideAd()
        {
            if (!IsInited)
            {
                Debug.WriteLine("HideAd called before initialization!");
                return;
            }

            if (!m_adShown)
            {
                Debug.WriteLine("HideAd called with no shown ads!");
                return;
            }

            try
            {
                m_adShown = false;
                AdWrapper wrapper = m_adWrappers[m_currentWrapper];
                wrapper.AdNetworkHandler.Hide();
                Debug.WriteLine("Called hide ad for network " + wrapper.AdNetworkHandler.Network);
            }
            catch (Exception ex)
            {
                Debug.WriteLine("HideAd Exception: " + ex);
            }
        }

        public void LoadAd()
        {
            if (!IsInited)
            {
                throw new ApplicationException("LoadAd called before initialization!");
            }

            m_currentWrapper++;
            m_adLoaded = false;

            Debug.WriteLine("Going to use wrapper #" + m_currentWrapper);

            if (m_currentWrapper >= m_adWrappers.Length)
            {
                Debug.WriteLine("No ad networks left for LoadAd()!");
                m_currentWrapper = -1;
                return;
            }

            AdWrapper wrapper = null;
            try
            {
                wrapper = m_adWrappers[m_currentWrapper];
                wrapper.CallCount++;

                wrapper.AdNetworkHandler.SetId(wrapper.Id);

                if (!wrapper.AdNetworkHandler.Show(wrapper, m_data))
                    throw new ApplicationException("Ad wrapper " + wrapper.Name + " returned false");

            }
            catch (Exception ex)
            {
                Debug.WriteLine("LoadAd Exception: " + ex);

                AdError(wrapper, ex.Message);
            }
        }

        public void DisplayAd()
        {
            if (!IsInited)
            {
                throw new ApplicationException("LoadAd called before initialization!");
            }

            if (!m_adLoaded)
            {
                Debug.WriteLine("DisplayAd called when no ad is loaded!");
                return;
            }

            if (m_currentWrapper == -1)
                LoadAd();

            if (m_currentWrapper == -1)
                return;

            AdWrapper wrapper = null;
            try
            {
                wrapper = m_adWrappers[m_currentWrapper];
                wrapper.AdNetworkHandler.Display();
            }
            catch (Exception ex)
            {
                Debug.WriteLine("DisplayAd Exception: " + ex);

                AdError(wrapper, ex.Message);
            }
        }

        internal void AdLoaded(AdWrapper wrapper)
        {
            string networkName = "Unknown network";
            try
            {
                m_adLoaded = true;

                if (wrapper != null)
                {
                    networkName = wrapper.Name;
                    wrapper.SuccessCount++;
                    Debug.WriteLine("Loaded ad from {0} [{1}/{2}/{3}/{4}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount);
#if VERBOSE_LOG
                    m_reportManager.ReportEvent("AD_LOAD", string.Format("{0} [{1}/{2}/{3}/{4}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount));
#endif
                }
                else
                {
                    Debug.WriteLine("Loaded ad from " + networkName);
#if VERBOSE_LOG
                    m_reportManager.ReportEvent("AD_LOAD", "");
#endif
                }

            }
            finally
            {
                m_context.OnLoadedAd(this, networkName);
            }
        }

        internal void AdError(AdWrapper wrapper, string error)
        {
            string networkName = "Unknown network";
            try
            {
                m_adShown = false;
                m_adLoaded = false;

                if (wrapper != null)
                {
                    networkName = wrapper.Name;
                    wrapper.LastError = error;
                    wrapper.FailCount++;
                    Debug.WriteLine("Failed to load ad from {0} [{1}/{2}/{3}/{4}], error {5}", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount, error);
                    m_reportManager.ReportEvent("AD_ERROR", string.Format("{0} [{1}/{2}/{3}/{4}] '{5}'", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount, error));
                }
                else
                {
                    Debug.WriteLine("Failed to load ad from " + networkName + ", error " + error);
                    m_reportManager.ReportEvent("AD_ERROR", string.Format("{0} '{1}'", networkName, error));
                }

            }
            finally
            {
                if (m_context.OnFailedAd(this, networkName, error))
                    LoadAd();
            }
        }

        internal void AdClicked(AdWrapper wrapper)
        {
            string networkName = "Unknown network";
            try
            {
                if (wrapper != null)
                {
                    networkName = wrapper.Name;
                    wrapper.ClickCount++;
                    Debug.WriteLine("Clicked ad from {0} [{1}/{2}/{3}/{4}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount);
                    m_reportManager.ReportEvent("AD_CLICK", string.Format("{0} [{1}/{2}/{3}/{4}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount));
                }
                else
                {
                    Debug.WriteLine("Clicked ad from " + networkName);
                    m_reportManager.ReportEvent("AD_CLICK", networkName);
                }
            }
            finally
            {
                m_context.OnClickedAd(this, networkName);
            }
        }

        internal void AdShown(AdWrapper wrapper, object view)
        {
            string networkName = "Unknown network";
            try
            {
                m_adShown = true;

                if (wrapper != null)
                {
                    networkName = wrapper.Name;
                    wrapper.ShowCount++;
                    Debug.WriteLine("Shown ad from {0} [{1}/{2}/{3}/{4}/{5}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount, wrapper.ShowCount);
#if VERBOSE_LOG
                    m_reportManager.ReportEvent("AD_SHOW", string.Format("{0} [{1}/{2}/{3}/{4}/{5}]", networkName, wrapper.CallCount, wrapper.SuccessCount, wrapper.FailCount, wrapper.ClickCount, wrapper.ShowCount));
#endif                
                }
                else
                {
                    Debug.WriteLine("Shown ad from " + networkName);
#if VERBOSE_LOG
                    m_reportManager.ReportEvent("AD_SHOW", networkName);
#endif
                }
            }
            finally
            {
                m_context.OnShownAd(this, networkName, view);
            }
        }
    }
}

File: Apps4Bro.Swift/Sources/Apps4Bro/AdManager.swift
import Foundation
import UIKit

// Port of AdManager.cs. The manager owns the network registry, fetches the
// ordered "try this network, fall back to that one" config from the App4Bro
// backend, and dispatches the public `loadAd`/`displayAd`/`hideAd` calls.
public final class AdManager {
    private static let app4BroTag = "app4bro"
    private static let TAG = "AdManager" // matches Android Log.d(TAG, ...) prefix

    private let reportManager: ReportManager
    private let session: URLSession

    private var adNetworks: [String: AdNetworkHandler] = [:]
    private let keys: [String: String]
    /// App number (e.g. "5" for Klondike). Exposed `internal` so network handlers
    /// can include it in HouseAdUrl/etc. requests for consistent analytics labels.
    let appId: String

    private var adWrappers: [AdWrapper]?
    private var currentWrapper: Int = -1

    private(set) public var isAdShown: Bool = false
    private(set) public var isAdLoaded: Bool = false
    private var inited: Bool = false

    public weak var context: AdContextDelegate?

    public var isInited: Bool { adWrappers != nil }

    /// Pass keys as `["app4bro", "<your-app-id>", "admob", "<unit-id>", ...]` to mirror
    /// the C# `params string[] keys` API. The first match for "app4bro" becomes the appId.
    public init(keys: [String]) {
        var dict: [String: String] = [:]
        for i in stride(from: 0, to: keys.count - 1, by: 2) {
            dict[keys[i].lowercased()] = keys[i + 1]
        }
        guard let appId = dict[Self.app4BroTag] else {
            Log.e(Self.TAG, "No App4Bro key specified for AdManager!")
            fatalError("No App4Bro key specified for AdManager!")
        }
        self.keys = dict
        self.appId = appId
        self.reportManager = ReportManager(applicationId: appId)

        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 5
        self.session = URLSession(configuration: cfg)

        Log.d(Self.TAG, "AdManager inited with id \(appId)")
        reportManager.reportEvent("AD_START", parameter: String(Apps4BroSDK.version))

        // Auto-register the built-in networks; host can still call `registerAdNetwork`
        // for custom ones before calling `initialize(context:)`. Order matters: the
        // wrapper list returned by the server picks among these by name.
        registerAdNetwork(AdMobNetwork(manager: self))
        registerAdNetwork(AdMobBannerNetwork(manager: self))
        registerAdNetwork(HouseNetwork(manager: self))
        registerAdNetwork(HouseInterNetwork(manager: self))
        registerAdNetwork(DummyNetwork(manager: self))

        preInit()
    }

    public func registerAdNetwork(_ handler: AdNetworkHandler) {
        adNetworks[handler.network.lowercased()] = handler
    }

    /// C# Init(context) — once the network fetch finishes, OnInited fires on the delegate.
    public func initialize(context: AdContextDelegate) {
        if self.context == nil && inited {
            self.context = context
            context.onInited(self)
            return
        }
        if isInited {
            loadAd()
            return
        }
        self.context = context
    }

    // MARK: - Server fetch

    private func formatAdRequest() -> String {
        // New args (App4Bro ID in path + id=; analytics name in app=; did= trails).
        // did= may be empty when ATT denied / SDK not inited.
        let analyticsName = Bundle.main.bundleIdentifier ?? ""
        let lang = Locale.current.languageCode ?? "en"
        return String(format: Apps4BroSDK.adManagerUrl,
                      appId,
                      analyticsName,
                      lang,
                      Apps4BroSDK.version,
                      Apps4BroSDK.platform,
                      Apps4BroSDK.advertisingId)
    }

    private func preInit() {
        let urlString = formatAdRequest()
        Log.d(Self.TAG, "We are starting requestAds: \(urlString)")
        guard let url = URL(string: urlString) else {
            Log.w(Self.TAG, "Bad URL: \(urlString)")
            DispatchQueue.main.async { self.doInit(data: nil, ok: false) }
            return
        }
        var req = URLRequest(url: url)
        req.timeoutInterval = 5
        session.dataTask(with: req) { [weak self] data, response, error in
            DispatchQueue.main.async {
                guard let self = self else { return }
                let body = data.flatMap { String(data: $0, encoding: .utf8) }
                let ok = error == nil && (response as? HTTPURLResponse)?.statusCode == 200
                if let err = error {
                    Log.w(Self.TAG, "Network error: \(err)")
                } else if let body = body, !body.isEmpty {
                    Log.d(Self.TAG, "Got response: \(body)")
                } else {
                    Log.w(Self.TAG, "Got empty response")
                }
                self.doInit(data: body, ok: ok)
            }
        }.resume()
    }

    private func doInit(data: String?, ok: Bool) {
        var fromCache = false
        var fromFallback = false
        if let body = data, !body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            parseAdData(body, save: true)
        }
        // Cached fallback from a previous good fetch.
        if adWrappers == nil {
            if let cached = UserDefaults.standard.string(forKey: Self.app4BroTag + "_" + appId) {
                Log.d(Self.TAG, "Using cached ad config")
                fromCache = true
                parseAdData(cached, save: false)
            }
        }
        // Hard-coded fallback from the constructor keys (each non-app4bro key becomes its own wrapper).
        if adWrappers == nil {
            Log.d(Self.TAG, "Registering fallback ads")
            fromFallback = true
            var builder = ""
            var count = 0
            for (k, v) in keys where k != Self.app4BroTag {
                count += 1
                builder += "\(k):\(k)def\(count)|\(v)|"
            }
            parseAdData(builder, save: false)
            if adWrappers == nil {
                Log.w(Self.TAG, "doInit: no wrappers after server/cache/fallback — staying uninitialized")
                return
            }
        }
        currentWrapper = -1
        inited = true
        let source = fromCache ? " (from cache)" : (fromFallback ? " (from fallback keys)" : "")
        Log.d(Self.TAG, "Initializing complete. Networks registered: \(adWrappers?.count ?? 0)\(source)")
        context?.onInited(self)
    }

    /// Parses the server response format `network1:zone1|adUnit1|network2:zone2|adUnit2|...`.
    private func parseAdData(_ data: String, save: Bool) {
        let parts = data.split(separator: "|", omittingEmptySubsequences: false).map(String.init)
        guard parts.count > 1 else { return }

        Log.d(Self.TAG, "Got \(parts.count / 2) items from server")

        var wrappers: [AdWrapper] = []
        var i = 0
        while i + 1 < parts.count {
            var networkSpec = parts[i].lowercased()
            let zoneId: String
            if networkSpec.contains(":") {
                let split = networkSpec.split(separator: ":", maxSplits: 1).map(String.init)
                networkSpec = split[0]
                zoneId = split[1]
            } else {
                zoneId = "\(networkSpec) \(wrappers.count + 1)"
            }
            if let handler = adNetworks[networkSpec] {
                wrappers.append(AdWrapper(handler: handler, id: parts[i + 1], name: zoneId))
                Log.d(Self.TAG, "Registered ad network \(networkSpec)[\(parts[i + 1])]")
            } else {
                Log.w(Self.TAG, "Failed to register ad network \(networkSpec)")
            }
            i += 2
        }
        guard !wrappers.isEmpty else { return }
        adWrappers = wrappers
        if save {
            UserDefaults.standard.set(data, forKey: Self.app4BroTag + "_" + appId)
            Log.d(Self.TAG, "Cache refreshed for next launch")
        }
        reportManager.reportEvent("AD_INIT", parameter: String(wrappers.count))
    }

    // MARK: - Public ad-flow API

    public func hideAd() {
        if !isInited {
            Log.w(Self.TAG, "hideAd called before initialization!")
            return
        }
        guard isAdShown else { return }
        isAdShown = false
        guard let wrappers = adWrappers, currentWrapper >= 0, currentWrapper < wrappers.count else { return }
        Log.d(Self.TAG, "hideAd: hiding wrapper \(wrappers[currentWrapper].name)")
        wrappers[currentWrapper].handler.hide()
    }

    public func loadAd() {
        guard isInited, let wrappers = adWrappers else {
            Log.e(Self.TAG, "loadAd called before initialization!")
            fatalError("loadAd called before initialization!")
        }
        if isAdShown {
            Log.w(Self.TAG, "loadAd called while showing")
        }
        currentWrapper += 1
        isAdLoaded = false
        if currentWrapper >= wrappers.count {
            Log.w(Self.TAG, "loadAd: exhausted all \(wrappers.count) wrappers, giving up")
            currentWrapper = -1
            return
        }
        let wrapper = wrappers[currentWrapper]
        wrapper.callCount += 1
        wrapper.handler.setId(wrapper.id)
        Log.d(Self.TAG, "loadAd: trying wrapper \(wrapper.name) (\(currentWrapper + 1)/\(wrappers.count))")
        if !wrapper.handler.show(data: wrapper) {
            Log.w(Self.TAG, "loadAd: wrapper \(wrapper.name) returned false synchronously")
            adError(data: wrapper, error: "Ad wrapper \(wrapper.name) returned false")
        }
    }

    public func displayAd() {
        guard isInited else {
            Log.e(Self.TAG, "displayAd called before initialization!")
            fatalError("displayAd called before initialization!")
        }
        guard isAdLoaded, let wrappers = adWrappers, currentWrapper >= 0, currentWrapper < wrappers.count else {
            Log.d(Self.TAG, "displayAd: no ad ready (loaded=\(isAdLoaded), idx=\(currentWrapper))")
            return
        }
        Log.d(Self.TAG, "displayAd: presenting wrapper \(wrappers[currentWrapper].name)")
        wrappers[currentWrapper].handler.display()
    }

    // MARK: - Network callbacks (called by AdNetworkHandler implementations)

    func adLoaded(data: Any?) {
        var networkName = "Unknown network"
        defer { context?.onLoadedAd(self, network: networkName) }
        isAdLoaded = true
        if let w = data as? AdWrapper {
            networkName = w.name
            w.successCount += 1
            Log.d(Self.TAG, "Ad loaded: \(stats(prefix: networkName, w: w))")
            reportManager.reportEvent("AD_LOAD", parameter: stats(prefix: networkName, w: w))
        } else {
            Log.d(Self.TAG, "Ad loaded: \(networkName)")
            reportManager.reportEvent("AD_LOAD", parameter: "")
        }
    }

    func adError(data: Any?, error: String) {
        var networkName = "Unknown network"
        defer {
            if context?.onFailedAd(self, network: networkName, error: error) == true {
                loadAd()
            }
        }
        isAdShown = false
        isAdLoaded = false
        if let w = data as? AdWrapper {
            networkName = w.name
            w.lastError = error
            w.failCount += 1
            Log.w(Self.TAG, "Ad error: \(stats(prefix: networkName, w: w)) '\(error)'")
            reportManager.reportEvent("AD_ERROR", parameter: "\(stats(prefix: networkName, w: w)) '\(error)'")
        } else {
            Log.w(Self.TAG, "Ad error: \(networkName) '\(error)'")
            reportManager.reportEvent("AD_ERROR", parameter: "\(networkName) '\(error)'")
        }
    }

    func adClicked(data: Any?) {
        var networkName = "Unknown network"
        defer { context?.onClickedAd(self, network: networkName) }
        if let w = data as? AdWrapper {
            networkName = w.name
            w.clickCount += 1
            Log.d(Self.TAG, "Ad clicked: \(stats(prefix: networkName, w: w))")
            reportManager.reportEvent("AD_CLICK", parameter: stats(prefix: networkName, w: w))
        } else {
            Log.d(Self.TAG, "Ad clicked: \(networkName)")
            reportManager.reportEvent("AD_CLICK", parameter: networkName)
        }
    }

    func adShown(data: Any?, view: Any?) {
        var networkName = "Unknown network"
        defer { context?.onShownAd(self, network: networkName, view: view) }
        isAdShown = true
        if let w = data as? AdWrapper {
            networkName = w.name
            w.showCount += 1
            let statsStr = "\(networkName) [\(w.callCount)/\(w.successCount)/\(w.failCount)/\(w.clickCount)/\(w.showCount)]"
            Log.d(Self.TAG, "Ad shown: \(statsStr)")
            reportManager.reportEvent("AD_SHOW", parameter: statsStr)
        } else {
            Log.d(Self.TAG, "Ad shown: \(networkName)")
            reportManager.reportEvent("AD_SHOW", parameter: networkName)
        }
    }

    private func stats(prefix: String, w: AdWrapper) -> String {
        "\(prefix) [\(w.callCount)/\(w.successCount)/\(w.failCount)/\(w.clickCount)]"
    }
}
